### PR TITLE
state_capturer: pin the exact host-cache size via mmap + cudaHostRegister

### PR DESCRIPTION
**Pull request by yueming-yuan**
BaseHostCache sizes its buffer by the KV pool's token capacity, and torch's caching host allocator rounds pinned blocks up to the next power of two — on DeepSeek-V4 at 21M pool tokens the 20.4 GiB request pinned 32 GiB per rank (verified by intercepting the underlying `cudaHostAlloc`, which was exactly 2^35 bytes), i.e. 46 GB of dead host memory per 4-GPU GB300 node for `--use-rollout-routing-replay`. Allocate through the same mmap + cudaHostRegister path the HiCache host pools use (`pool_host.common.ALLOC_MEMORY_FUNCS`), which pins exactly the requested size; npu/musa keep the plain `pin_memory=True` path via the existing dispatch.

Verified on GB300: exact nbytes (zero rounding), `is_pinned()` true, zero-init preserved, D2H/H2D roundtrip intact, `torch.cuda.host_memory_stats()` shows zero caching-host-allocator segments.

Port of #36562 (merged into `sglang-miles`) to main.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #33813337093](https://github.com/sgl-project/sglang/actions/runs/33813337093)
Latest PR Test (Extra): :x: [Run #33813336374](https://github.com/sgl-project/sglang/actions/runs/33813336374)
Latest PR Test (AMD ROCm 7.2): :no_entry_sign: [Run #33813336992](https://github.com/sgl-project/sglang/actions/runs/33813336992)

### CODE REVIEW

File: python/sglang/srt/state_capturer/base.py [hunk 1]
```diff
@@ -52,13 +56,18 @@ def _log_allocation(self):
 
 
 class BaseHostCache:
-    def __init__(self, num_tokens: int, num_layers: int, topk_size: int, name: str):
-        self.buffer = torch.zeros(
+    def __init__(
+        self, num_tokens: int, num_layers: int, topk_size: int, name: str, device: str
+    ):
+        alloc = ALLOC_MEMORY_FUNCS[device]
+        self.buffer = alloc(
```
Comment (fzyzcjy) [inline python/sglang/srt/state_capturer/base.py:68]: **Codex review (P2): Unregister the mmap-backed cache during shutdown.**

This is the same externally registered allocation path that HiCache already tears down explicitly, not merely a similar kind of pinned allocation.

**State-capturer allocation path introduced here**

- `BaseTopkCapturer.__init__` owns a `BaseHostCache`, and both `RoutedExpertsCapturer` and `IndexerTopkCapturer` derive from it.
- `BaseHostCache.__init__` selects `ALLOC_MEMORY_FUNCS[device]` and passes both `pin_memory=True` and `HostTensorAllocator()`.
- For CUDA, `ALLOC_MEMORY_FUNCS` has no explicit `"cuda"` entry, so its `defaultdict` selects `alloc_with_host_register`.
- `alloc_with_host_register` first calls `HostTensorAllocator.allocate`, which calls `alloc_mmap`, and then calls `_cuda_host_register` because `pin_memory=True`.
- `_cuda_host_register` calls `torch.cuda.cudart().cudaHostRegister(...)` and stores every registered `(ptr, size)` range on the tensor for later paired unregistration.
- The concrete call chain is therefore `BaseHostCache.__init__` -> `ALLOC_MEMORY_FUNCS["cuda"]` -> `alloc_with_host_register` -> `HostTensorAllocator.allocate` -> `alloc_mmap` -> `_cuda_host_register` -> `cudaHostRegister`.

**Existing HiCache allocation and cleanup path**

- `HostKVCache.__init__` creates its allocator with `get_allocator_from_storage`; the default allocator is the same `HostTensorAllocator`.
- Its subclass `init_kv_buffer` implementations also select `ALLOC_MEMORY_FUNCS[self.device_pool.device]` and pass `pin_memory=self.pin_memory` plus that allocator. On CUDA with pinning enabled, this reaches the same `alloc_with_host_register` -> `_cuda_host_register` path above.
- `HostKVCache.destroy` explicitly iterates its registered buffers and calls the same module's `_cuda_host_unregister` on CUDA/HIP before dropping `kv_buffer`. The method is idempotent and documents that leaving large registrations for kernel process reclaim can stall teardown for tens of seconds.
- On graceful exit, `run_scheduler_process` calls `Scheduler.release_host_resources()`, which calls `tree_cache.release_host_resources()`. The HiCache implementations then call the host pool/group `destroy()`, eventually reaching `HostKVCache.destroy()` -> `_cuda_host_unregister()` -> `cudaHostUnregister(...)`.

**Missing ownership edge for this PR**

- `ModelRunner.init_routed_experts_capturer` and `init_indexer_capturer` install the two capturers into `RuntimeResources.experts_capturer` and `RuntimeResources.indexer_capturer`.
- `BaseHostCache` and `BaseTopkCapturer` provide no `destroy()` method, so neither owner can release the registration explicitly.
- `Scheduler.release_host_resources()` releases HiCache, HiSparse, and decode-offload host resources, but it never destroys or clears either global state capturer.
- Consequently, these mmap-backed buffers remain CUDA-registered until process reclaim even on the graceful path. For the DeepSeek-V4 case described by this PR, that is roughly 20 GiB per rank.

Please mirror `HostKVCache`'s guarded, idempotent lifecycle: add a paired `_cuda_host_unregister` path for `BaseHostCache`, expose it through the capturer, and invoke it for both global capturers from graceful shutdown before clearing their global references.

**Review by fzyzcjy — APPROVED** [review]

**Review by hnyls2002 — APPROVED** [review]

**Comment by yueming-yuan** [comment]
/rerun-test test/registered/rl/test_return_routed_experts.py test/registered/rl/test_return_indexer_topk.py
